src/components/organisms/ChatDrawer/ChatDrawer.tsx

```diff
@@ -97,6 +97,18 @@ const ChatDrawer: React.FC<PropsType> = ({
         .sort(chatSort),
     [chats, roomName, HIDE_BEFORE]
   );
+  let noofchats : number = chatsToDisplay? chatsToDisplay.length:0;
+
+  if(typeof window.localStorage.noofchats =='undefined'){
+    window.localStorage.noofchats=noofchats;
+  }
+  let notificationAlert=0;
+  console.log(window.localStorage.noofchats);
+  console.log(noofchats);
+  if(parseInt(window.localStorage.noofchats) !== noofchats){
+    notificationAlert=1;
+  }
+  window.localStorage.noofchats=noofchats;
```

Comment:
@jigar-jstechno Thinking about it a bit more, I think your intent here was to use `localStorage` to 'cache' the last count of chats, so that we can check to see if it has changed, and if so, show the notification 'badge'. Is that right?

Since the code you have here is just in the 'main' section of the component, it will be executed every single time this component is rendered/re-rendered (which will be a lot). The general way we would write code that needs to 'only run sometimes', or to store a previous value for future comparison is with React hooks such as `useEffect`, `useMemo`, `useState` or similar (see below for more details)

Instead of storing the 'last count of chats' and then seeing if it has changed, a better approach would be to check how many unread chats we currently have directly.

Looking at the 'type' of `chatsToDisplay`, it appears to be an array of `WithId<RestrictedChatMessage>[] | undefined`:

![image](https://user-images.githubusercontent.com/753891/107590619-ab337780-6c5c-11eb-9984-23f176e73a28.png)

`RestrictedChatMessage` is defined in:

- https://github.com/sparkletown/sparkle/blob/staging/src/store/actions/Chat.ts#L71-L73

The types here are a little convoluted/hard to read, but essentially:

- `RestrictedChatMessage` 'extends' from `NonGlobalChatMessage`, which 'extends' from `BaseChatMessage`

```typescript
export type BaseChatMessage = {
  from: string;
  text: string;
  ts_utc: firebase.firestore.Timestamp;
  deleted?: boolean;
};

type NonGlobalChatMessage = BaseChatMessage & {
  to: string;
};

export type RestrictedChatMessage = NonGlobalChatMessage & {
  type: ChatMessageType.Table | ChatMessageType.Room;
};
```

So the 'eventual shape' of a `RestrictedChatMessage` is something like:

```typescript
{
  type: ChatMessageType.Table | ChatMessageType.Room;
  to: string;
  from: string;
  text: string;
  ts_utc: firebase.firestore.Timestamp;
  deleted?: boolean;
}
```

Weirdly, it seems we don't actually have the same `isRead` field that we do on `PrivateChatMessage` (cc // @mike-lvov FYI), presumably because all of these messages are stored in the same 'global collection' in firestore, so it wouldn't be possible/efficient to store the `isRead` field on them (since then it would apply to everyone, and not just the individual user)

Looking back at https://github.com/sparkletown/internal-sparkle-issues/issues/179, it seems like the request was for this feature to be implemented for 'direct messages', so I believe the intent is that it only applies to 'private chats', not 'room chats'/similar (which is where you're making these changes at the moment)

Since `PrivateChatMessage` does have the `isRead` property, this will be a lot simpler to implement:

- https://github.com/sparkletown/sparkle/blob/staging/src/store/actions/Chat.ts#L75-L78

`PrivateChatMessage` 'extends' from `NonGlobalChatMessage`, which 'extends' from `BaseChatMessage`; so the 'eventual shape' of a `PrivateChatMessage` is something like:

```typescript
{
  type: ChatMessageType.Private;
  to: string;
  from: string;
  text: string;
  ts_utc: firebase.firestore.Timestamp;
  deleted?: boolean;
  isRead: boolean;
}
```

Searching for 'privatechats', we can see that currently it exists in a number of places across the codebase (this is one of the things that @mike-lvov is working on resolving with his work-in-progress (WIP) changes in https://github.com/sparkletown/sparkle/pull/1074):

- https://github.com/sparkletown/sparkle/search?q=privatechats&type=code

Off the top of my head, there are 2 main 'locations' that private chats exist within the codebase at the moment (not including the refactors being made in https://github.com/sparkletown/sparkle/pull/1074):

- `NavBar`: https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/NavBar/NavBar.tsx
  - Which is shown at the top of most pages on the platform
- `Sidebar`: https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/Sidebar/Sidebar.tsx
  -  Which I believe is only used on the main 'partymap' template at the moment (but will be refactored to be used across all of our templates eventually, possibly as part of https://github.com/sparkletown/sparkle/pull/1074)

Looking at the `NavBar` component, there already seems to be a `numberOfUnreadMessages` variable, which calculates how many unread chats there are, and appears to already be used as part of the display:

- https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/NavBar/NavBar.tsx#L106-L112
- https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/NavBar/NavBar.tsx#L257-L273

Looking at the `Sidebar` component, there also seems to be some functionality that should be showing unread chats, namely the `hasUnreadMessages` variable, which looks as though it is used to display a notification, but doesn't seem to directly include the counts there:

- https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/Sidebar/Sidebar.tsx#L31
- https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/Sidebar/Sidebar.tsx#L67-L75

`hasUnreadMessages` uses the `unreadMessagesSelector` to read from our global redux state:

- https://github.com/sparkletown/sparkle/blob/staging/src/utils/selectors.ts#L146

It seems that this selector uses `.some()` to see if at least 1 element in an array passes the test:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

While that would have been efficient when we only cared whether there were unread messages or not, it wouldn't allow us to see how many unread messages there are.

Searching for `unreadMessagesSelector`, we can see that it only appears to be used by this one feature, so it shouldn't be a big change to be able to get the information we need from it:

- https://github.com/sparkletown/sparkle/search?q=unreadMessagesSelector

By changing `.some()` to `.filter()` we would be able to get the full list of unread private messages back, instead of just a boolean:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

This could then be used back in `Sidebar` instead of `hasUnreadMessages`:

- https://github.com/sparkletown/sparkle/blob/staging/src/components/molecules/Sidebar/Sidebar.tsx#L31

Since it would be returning an array of the unread messages now, we would want to do something like:

```typescript
 const unreadMessageCount = useSelector(unreadMessagesSelector).length;
```

Which could then be used in the JSX below to check whether to display the badge, and include the count of the unread messages.

It's possible that this same selector/code could be used in the `NavBar` component as well, to reduce the code duplication/complexity there, but i'm not 100% if that would work off the top of my head (I haven't tried it). If you aren't able to work that out on your own, @mike-lvov may be able to tell you more there.

Since it seems we currently have this 'private message notification count' code in at least 2 places, it might make sense to create a small new React component that can encapsulate this logic, and then add the audio sound effect to that, rather than duplicating the code in both locations.

---

We are generally able to use standard features within React when we want to 'cache' a value like this. Some of the more common methods (depending on our use case) are React hooks, and using their 'dependency' fields to ensure the hook only re-runs/'breaks memoisation' when it's dependencies have changed.

The 2 most common hooks we tend to use here are `useEffect` and `useMemo`:

- https://reactjs.org/docs/hooks-reference.html#useeffect
- https://reactjs.org/docs/hooks-reference.html#usememo

We will also use `useState` quite a bit, for storing data that we want to 'persist' between component re-renders:

- https://reactjs.org/docs/hooks-reference.html#usestate

If you're not already familiar with how to use React hooks, it's probably worth reading up about them a bit on the React docs website (no need to read all of these, but they are useful resources to understand one of the core patterns/things we use across the codebase):

- https://reactjs.org/docs/hooks-intro.html
- https://reactjs.org/docs/hooks-overview.html
- https://reactjs.org/docs/hooks-state.html
- https://reactjs.org/docs/hooks-effect.html
- https://reactjs.org/docs/hooks-rules.html
- https://reactjs.org/docs/hooks-custom.html
- https://reactjs.org/docs/hooks-reference.html
- https://reactjs.org/docs/hooks-faq.html

---

Depending on what kind of data we are storing, and if it needs to be accessible in other places in the app, we may also store it in our 'global state', which is built on Redux:

- https://redux.js.org/

While we aren't using Redux for a lot of things yet, it's probably something that we will be moving more and more of our global state/etc across to as we cleanup the old code, and build out new features.

In this case, I don't think it's the type of thing that would need to use Redux, but I figured I would mention it for your knowledge/future awareness.